pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -200,17 +220,95 @@ func (r *MachineSyncReconciler) Reconcile(ctx context.Context, req reconcile.Req
 	case machinev1beta1.MachineAuthorityClusterAPI:
 		return r.reconcileCAPIMachinetoMAPIMachine(ctx, capiMachine, mapiMachine)
 	case machinev1beta1.MachineAuthorityMigrating:
-		logger.Info("machine currently migrating", "machine", mapiMachine.GetName())
+		logger.Info("Machine currently migrating", "machine", mapiMachine.GetName())
 		return ctrl.Result{}, nil
 	default:
-		logger.Info("machine AuthoritativeAPI has unexpected value", "AuthoritativeAPI", mapiMachine.Status.AuthoritativeAPI)
+		logger.Info("Machine AuthoritativeAPI has unexpected value", "AuthoritativeAPI", mapiMachine.Status.AuthoritativeAPI)
 		return ctrl.Result{}, nil
 	}
 }
 
 // reconcileCAPIMachinetoMAPIMachine reconciles a CAPI Machine to a MAPI Machine.
-func (r *MachineSyncReconciler) reconcileCAPIMachinetoMAPIMachine(ctx context.Context, capiMachine *capiv1beta1.Machine, mapiMachine *machinev1beta1.Machine) (ctrl.Result, error) {
-	return ctrl.Result{}, nil
+func (r *MachineSyncReconciler) reconcileCAPIMachinetoMAPIMachine(ctx context.Context, capiMachine *capiv1beta1.Machine, mapiMachine *machinev1beta1.Machine) (ctrl.Result, error) { //nolint:funlen
+	logger := log.FromContext(ctx)
+
+	if capiMachine == nil {
+		logger.Error(errCAPIMachineNotFound, "machine", mapiMachine.Name)
+
+		if condErr := r.applySynchronizedConditionWithPatch(
+			ctx, mapiMachine, corev1.ConditionFalse, reasonCAPIMachineNotFound, errCAPIMachineNotFound.Error(), nil); condErr != nil {
+			return ctrl.Result{}, utilerrors.NewAggregate([]error{errCAPIMachineNotFound, condErr})
+		}
+
+		return ctrl.Result{}, errCAPIMachineNotFound
+	}
+
+	infraCluster, infraMachine, err := r.fetchCAPIInfraResources(ctx, capiMachine)
+	if err != nil {
+		fetchErr := fmt.Errorf("failed to fetch CAPI infra resources: %w", err)
+
+		if mapiMachine == nil {
+			r.Recorder.Event(capiMachine, corev1.EventTypeWarning, "SynchronizationWarning", fetchErr.Error())
+			return ctrl.Result{}, fetchErr
+		}
+
+		if condErr := r.applySynchronizedConditionWithPatch(
+			ctx, mapiMachine, corev1.ConditionFalse, reasonFailedToGetCAPIInfraResources, fetchErr.Error(), nil); condErr != nil {
+			return ctrl.Result{}, utilerrors.NewAggregate([]error{fetchErr, condErr})
+		}
+
+		return ctrl.Result{}, fetchErr
+	}
+
+	newMapiMachine, warns, err := r.convertCAPIToMAPIMachine(capiMachine, infraMachine, infraCluster)
+	if err != nil {
+		conversionErr := fmt.Errorf("failed to convert CAPI machine to MAPI machine: %w", err)
+
+		if mapiMachine == nil {
+			r.Recorder.Event(capiMachine, corev1.EventTypeWarning, "SynchronizationWarning", conversionErr.Error())
+			return ctrl.Result{}, conversionErr
+		}
```

Comment:
Nice! So we at least give the user some warning there!

Reply:
Yeah, that's the thinking. 

Do we need to update RBAC to allow events to be created on the CAPI machine? 🤔

Reply:
Probably not, it is already quite wide. We can try and see though.